shared/src/main/scala/coop/rchain/store/LazyAdHocKeyValueCache.scala

```diff
@@ -5,42 +5,81 @@ import cats.effect.Concurrent
 import cats.effect.concurrent.{Deferred, Ref}
 import cats.syntax.all._
 
+import scala.collection.immutable.LinearSeq
+
 trait KeyValueCache[F[_], K, V] {
   def get(key: K, fallback: F[V]): F[V]
   def put(key: K, value: V): F[Unit]
   def toMap: F[Map[K, V]]
 }
 
+trait InvalidationPolicy[V] {
+  def processNewInsert(item: V): (Option[V], InvalidationPolicy[V])
+}
+
+final class NoInvalidationPolicy[V] extends InvalidationPolicy[V] {
+  override def processNewInsert(item: V): (Option[V], InvalidationPolicy[V]) = (none[V], this)
+}
+
+final case class FixedSizeLRUInvalidationPolicy[V](
+    capacity: Long,
+    useOrder: LinearSeq[V] = LinearSeq()
+) extends InvalidationPolicy[V] {
+  def processNewInsert(item: V): (Option[V], InvalidationPolicy[V]) =
+    useOrder match {
+      case Seq() => (none[V], FixedSizeLRUInvalidationPolicy(capacity, useOrder))
+      case v     => (v.head.some, FixedSizeLRUInvalidationPolicy(capacity, useOrder.tail :+ item))
+    }
+}
+
+final case class KVCache[F[_], K, V](
+    items: Map[K, Deferred[F, V]],
+    purgePolicy: InvalidationPolicy[K]
+)
+
 final case class NoOpKeyValueCache[F[_]: Applicative, K, V]() extends KeyValueCache[F, K, V] {
   override def get(key: K, fallback: F[V]): F[V] = fallback
   override def put(key: K, value: V): F[Unit]    = ().pure[F]
   override def toMap: F[Map[K, V]]               = Map.empty[K, V].pure[F]
 }
 
 class LazyAdHocKeyValueCache[F[_]: Concurrent, K, V] private[LazyAdHocKeyValueCache] (
-    cache: Ref[F, Map[K, Deferred[F, V]]]
+    cache: Ref[F, KVCache[F, K, V]],
+    disableCache: Boolean
```

Comment:
The purpose of this class is to implement cache. It doesn't make sense to have a flag to disable it.
It should be responsibility of the consumer to not use if that is desirable.

Reply:
I agree, It's just the easiest way to do it. And this class actually not the cache itself - its a reader that caches everything passing through, Cache is not quite good name here. But I'm not sure yet whats the best way to handle this.